staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go

```diff
@@ -102,11 +102,18 @@ type ConfigFlags struct {
 	// before it is returned in ToRESTConfig function.
 	WrapConfigFn func(*rest.Config) *rest.Config
 
-	clientConfig clientcmd.ClientConfig
-	lock         sync.Mutex
-	// If set to true, will use persistent client config and
-	// propagate the config to the places that need it, rather than
-	// loading the config multiple times
+	clientConfig     clientcmd.ClientConfig
+	clientConfigLock sync.Mutex
+
+	restMapper     meta.RESTMapper
+	restMapperLock sync.Mutex
+
+	discoveryClient     discovery.CachedDiscoveryInterface
+	discoveryClientLock sync.Mutex
```

Comment:
Mutexes in Go and not re-entrant and discovery client calls `ToRESTConfig()` which holds the mutex, making it deadlock. Rest mapper calls discovery client getter so cannot use the same mutex too. It's possible to share the mutex by (un)locking multiple times, but it's cleaner to just use separate mutexes - cleaner code is easier to understand, maintain, and less chance of introducing a deadlock in the future.